group-coordinator/src/main/java/org/apache/kafka/coordinator/group/OffsetMetadataManager.java

```diff
@@ -180,7 +185,78 @@ public OffsetMetadataManager build() {
     /**
      * The offsets keyed by group id, topic name and partition id.
      */
-    private final TimelineHashMap<String, TimelineHashMap<String, TimelineHashMap<Integer, OffsetAndMetadata>>> offsetsByGroup;
+    private final Offsets offsets;
+
+    /**
+     * The offsets keyed by producer id, group id, topic name and partition id. This
```

Comment:
This comment about the keys is a little confusing.

Looking at the code we have 4 maps that are nested within each other. 😅 Maybe the comment could mention that each of these is a separate map? (Or we could include a non java doc comment in offsets to make it clearer.

Reply:
This comment also confused me because we also use the same nested 3 data structures to store the committed offsets it seems. So we have 

offsetsByGroup -> all committed offsets where we don't care about producer ID

pendingTransactionalOffsets -> a mapping of producer id -> offsetsByGroup (for only that producer) for all uncommitted txn offsets

On abort we remove the producer id's entries, and on commit we add them to offsetsByGroup. 

Is this correct?

Reply:
Sorry for so many comments here -- but is it correct we don't have the code to add/remove them on commit/abort yet?

Reply:
Let me rework those comments.

> On abort we remove the producer id's entries, and on commit we add them to offsetsByGroup.

Correct.

> Sorry for so many comments here -- but is it correct we don't have the code to add/remove them on commit/abort yet?

Correct. This will come in the next PR, stay tuned ;).